enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,570 @@
+
+In phase 3, deploying the WASP agent will not be needed.
+
+#### Workflow: Leveraging higher workload density
+
+1. The VM Owner is creating a regular virtual machine and is launching it.
+
+### API Extensions
+
+This proposal does not require any Kubernetes, OpenShift, or OpenShift
+Virtualization API changes.
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+The `MachineConfig` based swap provisioning will not work, as HCP does
+not provide the `MachineConfig` APIs.
+
+#### Standalone Clusters
+
+Standalone, regular and compact clusters are the primary use-cases for
+swap.
+
+#### Single-node Deployments or MicroShift
+
+Single-node and MicroShift deployments are out of scope of this
+proposal.
+
+### Implementation Details/Notes/Constraints
+
+#### WASP Agent
+
+At it's core the [wasp agent] is a `DaemonSet` delivering an [OCI Hook]
+which is used to turn on swap for selected workloads.
+
+Overall the agent is intended to align - and specifically not
+conflict - with the upstream Kubernetes SWAP design and behavior in
+order to simplify a transition.
+
+##### Design
+
+The design is driven by the following guiding principles:
+
+* System services are more important than workloads. Because workload
+  health depends on the health of system services.
+* Try to stay aligned to upstream Kubernetes swap behavior in order
+  to ease the transition to Kubernetes SWAP once available
+
+###### Enabling swap
+
+An OCI Hook to enable swap by setting the containers cgroup
+`memory.swap.max=max`.
+
+* **Technology Preview**
+  * Uses `UnlimitedSwap`.
+* **General Availability**
+  * Limited to burstable QoS class pods.
+  * Uses `LimitedSwap`.
+  * Limited to non-high-priority pods.
+
+For more info, refer to the upstream documentation on how to calculate
+[limited swap](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2400-node-swap#steps-to-calculate-swap-limit).
+
+###### Provisioning swap
+
+Provisioning of swap is left to the cluster administrator.
+The hook itself is not making any assumption where the swap is located.
+
+As long as there is no additional tooling available, the recommendation
+is to use `MachineConfig` objects to provision swap on nodes.
+
+The `MachineConfig` object would include the necessary scripts
+in order to provision swap on a disk, partition, or in a file.
+
+###### Node service protection
+
+All container workloads are run in the `kubepods.slice` cgroup.
+All system services are run in the `system.slice` cgroup.
+
+By default the `system.slice` is permitted to swap, however, system
+services are critical for a node's health and in turn critical for the
+workloads.
+
+Without system services such as `kubelet` or `crio`, any container will
+not be able to run well.
+
+Thus, in order to protect the `system.slice` and ensure that the nodes
+infrastructure health is prioritized over workload health, the agent is
+reconfiguring the `system.slice` and setting `memory.swap.max=0` to
```

Comment:
Reconfiguration is managed by MCO since we deliver this configuration via machine config. I think it could be raceful in case another MC will deployed, that will enable swapping for system slice.

Reply:
I believe that MCOs are ordered, so no race there. If you can say that the agent is configuring system.slice on boot time, it would be clearer.